File: xiequ_white_list.js
/*
new Env('携趣IP白名单');
51 23 * * * 
xiequ_white_list.js
by:Shuai
携趣自动加白名单
*/
const fs = require('fs');
const request = require('request');

// 配置UID和UKEY(api提取页面下面找)
let uid = process.env.XIEQU_UID || '';
let ukey = process.env.XIEQU_UKEY || '';

const ipFileName = 'xiequIp.txt';
const checkIpUrls = [
  'http://ip-api.com/json',
  'https://checkip.synology.com/',
  'http://httpbin.org/ip'
];

// 读取保存的IP
const readSavedIp = () => fs.existsSync(ipFileName) ? fs.readFileSync(ipFileName, 'utf8').trim() : null;

// 保存IP
const saveIp = (ip) => fs.writeFileSync(ipFileName, ip);

// 获取当前IP
const getCurrentIp = async () => {
  for (let url of checkIpUrls) {
    try {
      const currentIp = await new Promise((resolve, reject) => {
        request.get({ url, timeout: 5000 }, (err, res, body) => {
          if (err) {
            reject(err);
          } else {
            const match = body.match(/((25[0-5]|2[0-4]\d|[01]?\d\d?)\.){3}(25[0-5]|2[0-4]\d|[01]?\d\d?)/);
            resolve(match ? match[0] : null);
          }
        });
      });
      if (currentIp) {
        console.log(`🌐 当前IP: ${currentIp}`);
        return currentIp;
      }
    } catch (err) {
      console.log(`❗ 获取IP失败: ${err.code}`);
    }
  }
  return null;
};

// 添加IP到白名单
const addIpToWhiteList = async (ip) => {
  const url = `http://op.xiequ.cn/IpWhiteList.aspx?uid=${uid}&ukey=${ukey}&act=add&ip=${ip}`;
  request.get(url, (err, res, body) => {
    if (err) {
      console.log(`❗ 添加IP失败: ${err}`);
    } else if (body.includes('success')) {
      console.log(`✅ IP已添加到白名单: ${ip}`);
    } else {
      console.log(`❗ 添加IP失败: ${body}`);
    }
  });
};

// 删除白名单中的IP
const delWhiteIp = async (ip) => {
  const url = `http://op.xiequ.cn/IpWhiteList.aspx?uid=${uid}&ukey=${ukey}&act=del&ip=${ip}`;
  request.get(url, (err, res, body) => {
    if (err) {
      console.log(`❗ 删除IP失败: ${err}`);
    } else {
      console.log(`🗑️ 删除白名单中的IP: ${ip}`);
    }
  });
};

// 获取白名单IP列表
const getWhiteIpList = async () => {
  const url = `http://op.xiequ.cn/IpWhiteList.aspx?uid=${uid}&ukey=${ukey}&act=get`;
  return await new Promise((resolve, reject) => {
    request.get(url, (err, res, body) => {
      if (err) {
        reject(err);
      } else {
        resolve(body);
      }
    });
  });
};

// 主流程
const main = async () => {
  const currentIp = await getCurrentIp();
  if (!currentIp) {
    console.log('❗ 无法获取当前IP，终止执行！');
    return;
  }

  const savedIp = readSavedIp();
  const whiteIpList = await getWhiteIpList();

  if (savedIp && savedIp !== currentIp && whiteIpList.includes(savedIp)) {
    await delWhiteIp(savedIp);
  }

  if (!whiteIpList.includes(currentIp)) {
    await addIpToWhiteList(currentIp);
    saveIp(currentIp);
  } else {
    console.log('😎 当前IP已在白名单中，无需添加。');
  }
};

main();

File: xiequ_white_list.py
"""
new Env('携趣IP白名单');
45 23 * * * 
xiequ_white_list.py

by: https://github.com/Twshuai/white_list.git
携趣自动加白名单
"""

import requests

# 携趣代理的用户 ID 和密钥
UID = ''
UKEY = ''
BASE_URL = "http://op.xiequ.cn/IpWhiteList.aspx"

def get_public_ip():
    """获取公共 IP 地址"""
    response = requests.get('http://ip-api.com/json')
    if response.status_code == 200:
        data = response.json()
        if data['status'] == 'success':
            return data['query']
    return None

def get_whitelist():
    """获取当前白名单 IP 列表"""
    url = f"{BASE_URL}?uid={UID}&ukey={UKEY}&act=get"
    response = requests.get(url)
    
    if response.status_code == 200:
        print("📜 当前白名单 IP 列表：")
        print(response.text)
    else:
        print("❌ 获取白名单失败，状态码：", response.status_code)

def add_ip_to_whitelist(ip_address):
    """将指定 IP 添加到白名单"""
    url = f"{BASE_URL}?uid={UID}&ukey={UKEY}&act=add&ip={ip_address}"
    response = requests.get(url)
    
    if response.status_code == 200:
        if "Err:IpOutOfLimit" in response.text:
            print("⚠️ IP 已达到上限，无法添加。请删除一些 IP 后重试。")
        else:
            print("✅ IP 添加成功！响应内容：")
            print(response.text)
    else:
        print("❌ IP 添加失败，状态码：", response.status_code)

def delete_ip_from_whitelist(ip_address):
    """从白名单中删除指定 IP"""
    url = f"{BASE_URL}?uid={UID}&ukey={UKEY}&act=del&ip={ip_address}"
    response = requests.get(url)
    
    if response.status_code == 200:
        print(f"🗑️ IP {ip_address} 删除成功，响应内容：")
        print(response.text)
    else:
        print(f"❌ IP {ip_address} 删除失败，状态码：", response.status_code)

if __name__ == "__main__":
    # 获取当前的公共 IP 地址
    public_ip = get_public_ip()
    if public_ip:
        print(f"🌐 获取到的公共 IP 是: {public_ip}")
        
        # 获取当前的白名单 IP 列表
        get_whitelist()
        
        # 添加新的 IP 到白名单
        add_ip_to_whitelist(public_ip)
    else:
        print("⚠️ 无法获取公共 IP 地址")
